File: LeetCode/LeetCode/Array/ArrayAlgorithm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode.LeetCode.ArrayAlgorithm
{
    public class ArrayAlgorithm
    {
        public int LeetCode_704_Search(int[] nums, int target)
        {
            int left = 0;
            int right = nums.Length - 1;
            while (left <= right)
            {
                int mid = (left + right) / 2;
                if (nums[mid] == target)
                {
                    return mid;
                }
                else if (nums[mid] > target)
                {
                    right = mid - 1;
                }
                else if (nums[mid] < target)
                {
                    left = mid + 1;
                }
            }
            return -1;
        }
        public int LeetCode_27_RemoveElement(int[] nums, int val)
        {
            int slow = 0;
            for (int fast = 0; fast < nums.Length; fast++)
            {
                if (val != nums[fast])
                {
                    nums[slow++] = nums[fast];
                }
            }
            return slow;
        }
        public int[] LeetCode_977_SortedSquares(int[] nums)
        {
            //Method1  暴力解法
            for (int i = 0; i < nums.Length; i++)
            {
                nums[i] = nums[i] * nums[i];
            }

            Array.Sort(nums);
            return nums;

            //Method2 双指针
            int len = nums.Length;
            int[] res = new int[len];

            int k = len - 1;
            for (int i = 0,j=len-1; i <=j;)
            {
                if (nums[i] * nums[i] < nums[j] * nums[j])
                {
                    res[k--] = nums[j]*nums[j];
                    j--;
                }
                else
                {
                    res[k--] = nums[i] * nums[i];
                    i++;
                }
            }

            return res;
        }
        public int LeetCode_209_MinSubArrayLen(int target, int[] nums)
        {
            //Method1 暴力解法
            int res = int.MaxValue;
            int sum = 0;
            int len = nums.Length;
            //for (int i = 0; i < len; i++)
            //{
            //    sum = 0;
            //    for (int j = i; j < len; j++)
            //    {
            //        sum += nums[j];
            //        if (sum >= target)
            //        {
            //            res = Math.Min(res, j - i + 1);
            //            break; 
            //        } 
            //    }
            //}

            //return res == int.MaxValue ? 0 : res;

            //Method2 滑动窗口
            int index = 0;
            for (int i = 0; i < len; i++)
            {
                sum += nums[i];
                while (sum >= target)
                {
                    res = Math.Min(res, i - index + 1);
                    sum -= nums[index];
                    index++;
                }
            }

            return res == int.MaxValue ? 0 : res;

        }
        public int[][] LeetCode_59_GenerateMatrix(int n)
        {
            int[][] nums = new int[n][];
            for (int i = 0; i < n; i++)
            {
                nums[i] = new int[n];
            }
            int startX = 0,startY = 0;
            int offset = 1;
            int count = 1;
            int laps = n / 2;
            while (laps>0)
            {
                for (int i = startY; i < n - offset; i++)
                {
                    nums[startX][i] = count++;
                }

                for (int i = startX; i < n-offset; i++)
                {
                    nums[i][n-offset] = count++;
                }

                for (int i = n-offset; i > startX; i--)
                {
                    nums[n-offset][i] = count++;
                }

                for (int i = n-offset; i > startY; i--)
                {
                    nums[i][startY] = count++;
                }

                laps--;
                offset++;
                startX++;
                startY++;
            }

            if (n % 2 == 1)
            {
                nums[n / 2][n / 2] = n * n;
            }

            return nums;
        }
    }
}

File: LeedCodeCpp/ArrayAlgorithm.cpp
#include "ArrayAlgorithm.h"
#include <algorithm>

int ArrayAlgorithm::LeetCode704_search(vector<int>& nums, int target) {

	int i=0,j = nums.size()-1;
	while (i<=j)
	{
		int mid = i + (j - i) / 2;

		if (nums[mid] == target) {
			return mid;
		}
		else if (nums[mid] > target) {
			j = mid - 1;
		}
		else if (nums[mid] < target) {
			i = mid + 1;
		}
	}
	
	return -1;
}

int ArrayAlgorithm::Leetcode27_removeElement(vector<int>& nums, int val) {

	int left = 0, right = nums.size()- 1;
	
	while (left <= right) {
		if (nums[left] == val) {
			nums[left] = nums[right];
			right--;
		}
		else
		{
			left++;
		}
	}
	return left;

}

vector<int> ArrayAlgorithm::LeetCode_977_sortedSquares(vector<int>& nums) {

	//Method 1: 直接平方后排序
	//vector<int> result;
	//for (int num : nums)
	//{
	//	result.push_back(num * num);
	//}
	//sort(result.begin(), result.end());

	//Method 2: 双指针
	int len = nums.size();
	int left = 0;
	int right = len - 1;
	int pos = len - 1;
	vector<int> result(len);

	while (left <= right) {

		int leftSquare = nums[left] * nums[left];
		int rightSquare = nums[right] * nums[right];

		if (leftSquare > rightSquare) {
			result[pos--] = leftSquare;
			left++;
		}else {
			result[pos--] = rightSquare;
			right--;
		}
	}

	return result;
}

int ArrayAlgorithm::Leetcode209_minSubArrayLen(int target, vector<int>& nums) {
	
	// 滑动窗口
	int left = 0, right = 0;
	int minLength = INT_MAX;
	int sum = 0;
	int len = nums.size();
	while (right < len) {
		
		sum += nums[right];
		while(sum >= target) {
			minLength = min(minLength, right - left + 1);
			sum -= nums[left];
			left++;
		}
		right++;
	}
	
	return minLength == INT_MAX? 0 : minLength;
}

vector<vector<int>> ArrayAlgorithm::Leetcode59_generateMatrix(int n)
{
	vector<vector<int>> res(n,vector<int>(n,0));
	int top = 0, bottom = n - 1, left = 0, right = n - 1;
	int num = 1;
	while(num <= n*n) {
		
		for (int i = left;i <= right;i++) {
			res[top][i] = num++;
		}
		top++;
		for (int i = top; i <= bottom; i++)
		{
			res[i][right] = num++;
		}
		right--;
		for (int i = right; i >= left; i--)
		{
			res[bottom][i] = num++;
		}
		bottom--;
		for (int i = bottom; i >= top; i--)
		{
			res[i][left] = num++;
		}
		left++;
	}
	
	return res;
}

vector<int> ArrayAlgorithm::Leetcode54_spiralOrder(vector<vector<int>>& matrix)
{
	vector<int> res;
	int rows = matrix.size();
	int columns = matrix[0].size();
	int top = 0, bottom = matrix.size() - 1, left = 0, right = matrix[0].size() - 1;

	while (res.size() < rows*columns) {
		
		for (int i = left; i <= right && res.size() < rows * columns; i++) {

			res.push_back(matrix[top][i]);
		}
		top++;
		for (int i = top; i <=bottom && res.size() < rows * columns; i++)
		{
			res.push_back(matrix[i][right]);
		}
		right--;
		for (int i = right; i >= left && res.size() < rows * columns; i--)
		{
			res.push_back(matrix[bottom][i]);
		}
		bottom--;
		for (int i = bottom; i >= top && res.size() < rows * columns; i--)
		{
			res.push_back(matrix[i][left]);
		}
		left++;
	}

	return res;
}